bitcoin/src/pow.rs

```diff
@@ -913,6 +929,57 @@ impl kani::Arbitrary for U256 {
     }
 }
 
+// FIXME: we should move this in a better place
+/// TODO: add some docs here
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
+    last_height: Height,
+    last_time: u32,
+    last_target: Target,
+    adjustment_interval: u64,
+    pow_target_timespan: u64,
+    no_pow_retargeting: bool,
+    get_block_by_heigh: F,
+) -> u32
+where
+    F: Fn(u32) -> Header,
+{
+    let current_height: u64 = last_height.to_consensus_u32() as u64;
+    if (current_height % adjustment_interval.min(1)) != 0 {
+        return last_target.to_compact_lossy().to_consensus();
+    }
+
+    let last_adjustment_height = last_height.to_consensus_u32() as u64 - (adjustment_interval - 1);
+    let last_adjustment_block = get_block_by_heigh(last_adjustment_height as u32);
+    let last_adjustment_time = last_adjustment_block.time;
+
+    if no_pow_retargeting {
+        return last_adjustment_block.bits.to_consensus();
+    }
+
+    let actual_timespan = (last_time - last_adjustment_time) as u64;
+    let mut adjusted_timespan = actual_timespan;
+
+    if actual_timespan < pow_target_timespan / 4 {
+        adjusted_timespan = pow_target_timespan / 4;
+    }
+    if actual_timespan > pow_target_timespan * 4 {
+        adjusted_timespan = pow_target_timespan * 4;
+    }
```

Comment:
In 003673d1cb0495b80fcea5e01333eba9594d8e20:

I think it'd be clearer to early return in these cases. For the returned `U256` you can use shifts rather than multiplies which will be much more efficient.

Reply:
I agree, but this optimisation is not done but the compiler itself? it should be easy on a type u64 to eeeeeeeeeeedo this kind of optimization, is it?

Reply:
BTW in here, we can return early in both If? or just for the multiplication?

Reply:
In both cases. If we would adjust difficulty by a factor 4 in either direction, the adjustment is clamped by the factor 4.

Reply:
> In [003673d](https://github.com/rust-bitcoin/rust-bitcoin/commit/003673d1cb0495b80fcea5e01333eba9594d8e20):
> 
> I think it'd be clearer to early return in these cases. For the returned `U256` you can use shifts rather than multiplies which will be much more efficient.

I thought the compiler optimized this to a shift for us?

Reply:
> I agree, but this optimisation is not done but the compiler itself? 

I thought this would have been optimized by the compiler too. Anyone know for sure?

Reply:
@tcharding I'm sure it's optimized for primitives (`u8`-`u128` and `usize`) not for our `U256`. IIRC `pow_target_timespan` is `u32`.

Reply:
Even though the `U256` type is private? What if we used `target.0` to get at the `U256`?

Reply:
Oh of course, `U256` is an "arbitrary" type not an actual integer type. I was thinking it was a simple wrapper `U256(<integer type>)`.